src/lib.rs

```diff
@@ -175,41 +315,72 @@ pub enum Type {
     /// assert_eq!(format!("{}", &t), "(t0 → t1) → list(t0) → list(t1)");
     /// # }
     /// ```
-    Variable(u32),
+    Variable(Variable),
 }
 impl Type {
-    /// Whether a type has any type variables.
-    pub fn is_polymorphic(&self) -> bool {
-        match *self {
-            Type::Arrow(ref arr) => arr.arg.is_polymorphic() || arr.ret.is_polymorphic(),
-            Type::Constructed(_, ref args) => args.iter().any(|t| t.is_polymorphic()),
-            Type::Variable(_) => true,
+    /// shortcut for constructing function types (i.e. `alpha` → `beta`).
+    pub fn arrow(alpha: Type, beta: Type) -> Type {
+        Type::Constructed("→", vec![alpha, beta])
+    }
+    /// `true` if `Type` is an arrow type else `false`.
+    pub fn as_arrow(&self) -> Option<(&Type, &Type)> {
+        if let Type::Constructed("→", ref args) = *self {
+            Some((&args[0], &args[1]))
+        } else {
+            None
         }
     }
-    fn occurs(&self, v: u32) -> bool {
+    fn occurs(&self, v: Variable) -> bool {
         match *self {
-            Type::Arrow(ref arr) => arr.arg.occurs(v) || arr.ret.occurs(v),
             Type::Constructed(_, ref args) => args.iter().any(|t| t.occurs(v)),
             Type::Variable(n) => n == v,
         }
     }
-    /// Supplying is_return helps arrows look cleaner.
+    /// Supplying `is_return` helps arrows look cleaner.
     fn show(&self, is_return: bool) -> String {
         match *self {
-            Type::Arrow(ref arr) => arr.show(is_return),
+            Type::Variable(v) => format!("t{}", v),
             Type::Constructed(name, ref args) => {
                 if args.is_empty() {
                     String::from(name)
+                } else if name == "→" {
+                    Type::arrow_show(args, is_return)
                 } else {
                     format!("{}({})", name, args.iter().map(|t| t.show(true)).join(","))
                 }
             }
-            Type::Variable(v) => format!("t{}", v),
         }
     }
-    /// Applies the type in a context.
+    /// show specifically for arrow types
+    fn arrow_show(args: &[Type], is_return: bool) -> String {
+        if is_return {
+            format!("{} → {}", args[0].show(false), args[1].show(true))
+        } else {
+            format!("({} → {})", args[0].show(false), args[1].show(true))
+        }
+    }
+    /// Optionally return the arguments of an arrow.
+    pub fn args(&self) -> Option<VecDeque<&Type>> {
+        match *self {
+            Type::Constructed("→", ref args) => {
+                let mut tps = args[1].args().unwrap_or_default();
+                tps.push_front(&args[0]);
+                Some(tps)
+            }
+            Type::Variable(_) | Type::Constructed(..) => None,
+        }
+    }
+    /// Optionally return the return type of an arrow.
+    pub fn returns(&self) -> Option<&Type> {
```

Comment:
Maybe this and the preceding method (`args()`) should be replaced with something like:

```rust
/// Like [`as_arrow`], but recursively gets all curried argument types.
pub fn deconstruct_arrow(&self) -> Option<(VecDeque<&Type>, &Type)> { .. }
```

Reply:
Maybe, though my intuition is that I often have use for one or the other. Let's leave it for now and change later as use dictates.